backend/src/index.js

```diff
@@ -11,134 +11,110 @@ const app = require( 'express' )();
 const { Server } = require( 'http' );
 const http = new Server( app );
 const io = require( 'socket.io' )( http );
+const Redis = require( 'ioredis' );
+const redisAdapter = require( 'socket.io-redis' );
+io.adapter( redisAdapter( { host: 'localhost', port: 6379 } ) );
 
-const db = new Map();
+const client = new Redis();
 
 // Priorities are set from the lowest to the highest.
 const statePriorities = [ 'away', 'viewing', 'commenting', 'editing', 'merging' ];
 
 io.on( 'connection', socket => {
-	socket.session = {};
-
 	const timestamp = new Date();
 
-	socket.on( 'setUser', ( message, reply ) => {
-		const issueKey = `${ message.repoName }:${ message.issueId }`;
-		if ( !db.has( issueKey ) ) {
-			db.set( issueKey, new Map() );
-		}
+	socket.on( 'setUser', async ( message, reply ) => {
+		const issueKey = createIssueKey( message.repoName, message.issueId );
+		const issueSessions = await getSessionsFromDatabase( issueKey );
 
-		if ( !socket.session.issues ) {
-			socket.session.issues = [];
+		if ( !socket.issueKey ) {
+			socket.issueKey = issueKey;
 		}
 
-		if ( !socket.session.issues.includes( issueKey ) ) {
-			socket.session.issues.push( issueKey );
-		}
-
-		const issueUsers = db.get( issueKey );
-
-		if ( !issueUsers.has( message.user.login ) ) {
-			issueUsers.set( message.user.login, { sockets: [], tabs: new Map(), joinedAt: timestamp } );
+		if ( !issueSessions.hasOwnProperty( issueKey ) ) {
+			issueSessions[ socket.id ] = Object.assign( message.user, { joinedAt: timestamp } );
 		}
 
-		const issueUser = issueUsers.get( message.user.login );
-
-		if ( !issueUser.sockets.includes( socket.id ) ) {
-			socket.session.login = message.user.login;
+		const issueSession = issueSessions[ socket.id ];
 
+		if ( !issueSessions.hasOwnProperty( socket.id ) ) {
 			socket.join( issueKey );
-
-			issueUser.sockets.push( socket.id );
 		}
 
-		issueUser.tabs.set( socket.id, message.user.state );
-
-		issueUser.user = message.user;
+		const users = getUsers( issueSessions );
 
-		const users = getUsers( issueUsers );
+		await client.hset( issueKey, socket.id, JSON.stringify( issueSession ) );
```

Comment:
Can we have a similar method to save like the `getSessionsFromDatabase` for loading?

Reply:
I've rebuilt the project - now HangHub is using repository pattern and whole communication like getting, setting, deleting etc. is handled by the repository.